File: Design-Patterns-Using-CSharp/Prototype/PrototypePatternShape.cs
using System;
using System.Collections.Generic;

abstract class Shape : ICloneable
{
    protected string color;

    public Shape()
    {
        this.color = "Black";
    }

    public abstract override string ToString();

    public object Clone()
    {
        return MemberwiseClone();
    }

    public abstract Shape CloneShape();
}

class Rectangle : Shape
{
    public override string ToString()
    {
        return "Rectangle of colour " + color;
    }

    public override Shape CloneShape()
    {
        try
        {
            return (Shape)Clone();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
            return null;
        }
    }
}

class Circle : Shape
{
    public override string ToString()
    {
        return "Circle of colour " + color;
    }

    public override Shape CloneShape()
    {
        try
        {
            return (Shape)Clone();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
            return null;
        }
    }
}

class ShapeRegistry
{
    private static readonly Dictionary<string, Shape> Shapes = new Dictionary<string, Shape>();

    static ShapeRegistry()
    {
        Load();
    }

    public static void AddShape(string key, Shape value)
    {
        Shapes[key] = value;
    }

    public static Shape GetShape(string key)
    {
        if (Shapes.ContainsKey(key))
        {
            return Shapes[key].CloneShape();
        }
        return null;
    }

    public static void Load()
    {
        AddShape("circle", new Circle());
        AddShape("rectangle", new Rectangle());
    }
}

// Client code

public class PrototypePatternShape
{
    public static void Main(string[] args)
    {
        ShapeRegistry.Load();
        Shape c = ShapeRegistry.GetShape("circle");
        Shape r = ShapeRegistry.GetShape("rectangle");
        Console.WriteLine(c);
        Console.WriteLine(r);
    }
}

/*
Circle of colour Black
Rectangle of colour Black
 */

File: Design-Patterns-Using-Cpp/Prototype/PrototypePatternShape.cpp
#include <iostream>
#include <unordered_map>

class Shape {
protected:
    std::string color;

public:
    Shape() : color("") {}

    virtual ~Shape() {}

    virtual std::string toString() const = 0;

    virtual Shape* clone() const = 0;
};

class Rectangle : public Shape {
public:
    std::string toString() const override {
        return "Rectangle.";
    }

    Shape* clone() const override {
        return new Rectangle(*this);
    }
};

class Circle : public Shape {
public:
    std::string toString() const override {
        return "Circle.";
    }

    Shape* clone() const override {
        return new Circle(*this);
    }
};

class ShapeRegistry {
private:
    static std::unordered_map<std::string, Shape*> shapes;

public:
    static void addShape(const std::string& key, Shape* value) {
        shapes[key] = value;
    }

    static Shape* getShape(const std::string& key) {
        if (shapes.find(key) != shapes.end()) {
            return shapes[key]->clone();
        }
        return nullptr;
    }

    static void load() {
        addShape("circle", new Circle());
        addShape("rectangle", new Rectangle());
    }
};

std::unordered_map<std::string, Shape*> ShapeRegistry::shapes;

// Client code
int main() {
    ShapeRegistry::load();

    Shape* c = ShapeRegistry::getShape("circle");
    Shape* r = ShapeRegistry::getShape("rectangle");

    if (c && r) {
        std::cout << c->toString() << " " << r->toString() << std::endl;

        delete c;
        delete r;
    }

    return 0;
}

/*
Circle. Rectangle.
*/